src/main/java/com/ibm/cloudant/kafka/transforms/KeyToDocId.java

```diff
@@ -0,0 +1,146 @@
+
+public class KeyToDocId<R extends ConnectRecord<R>> implements Transformation<R> {
+
+    public static final String OVERVIEW_DOC = "Extract the record's key and save the value to the '_id' field. " +
+            "If the '_id' record already exists, the value will be overwritten." +
+            "If the record key doesn't exist, the record will be unmodified.";
+
+    public static final ConfigDef CONFIG_DEF = new ConfigDef();
+
+    private static final String PURPOSE = "Copy record key to _id value";
+
+    private static final String idFieldName = "_id";
+
+    private Cache<Schema, Schema> schemaUpdateCache = new SynchronizedCache<>(new LRUCache<Schema, Schema>(16));
+
+    private static Logger LOG = LoggerFactory.getLogger(KeyToDocId.class.toString());
+
+    public void configure(Map<String, ?> configs) {
+        // no config required
+    }
+
+    public R apply(R record) {
+        if (record.key() != null
+                && String.valueOf(record.key()).length() > 0) {
+            LOG.info("Attempting to apply KeyToDocId transform");
+            if (record.valueSchema() == null) {
+                return applySchemaless(record);
+            } else {
+                return applyWithSchema(record);
+            }
+        } else {
+            LOG.info("The topic message does not have a key. Leaving message unmodified.");
+           return record;
+        }
+    }
+
+    private R applySchemaless(R record) {
+        final Map<String, Object> value;
+        if (record.value() == null) {
+            value = new HashMap<>(1);
+        } else {
+            value = requireMapOrNull(record.value(), PURPOSE);
+        }
+        final Map<String, Object> updatedValue = new HashMap<>(value);
+        String keyToUseAsId = String.valueOf(record.key());
+        if (keyToUseAsId != null) {
+            updatedValue.put(idFieldName, keyToUseAsId);
+        } else {
+            LOG.info("The topic message does not have a key. Leaving message unmodified.");
+        }
+        return record.newRecord(
+                record.topic(),
+                record.kafkaPartition(),
+                record.keySchema(),
+                record.key(),
+                record.valueSchema(),
+                updatedValue,
+                record.timestamp()
+        );
+    }
+
+    private R applyWithSchema(R record) {
+        final Struct value = requireStruct(record.value(), PURPOSE);
+
+        Schema updatedSchema = schemaUpdateCache.get(value.schema());
+        if (updatedSchema == null) {
+            updatedSchema = makeUpdatedSchema(value.schema());
+            schemaUpdateCache.put(value.schema(), updatedSchema);
+        }
+
+        final Struct updatedValue = new Struct(updatedSchema);
+        for (Field field : value.schema().fields()) {
+            updatedValue.put(field.name(), value.get(field));
+        }
+
+        String keyToUseAsId = String.valueOf(record.key());
+        if (!keyToUseAsId.equals("Struct{}")) {
```

Comment:
Yea, I recognized it's not a great check but couldn't figure out a good way to check that the key was actually empty.  The Struct class has no way to check if values exists or the size of all values.

Reply:
`((Struct) record.key()).schema().fields().isEmpty()`

Obviously you have to already know the key is a `Struct` to do that. That is part of the reason I think we should do the key handling in one place so we can convert it to an appropriate type and do sensible checks. Obviously that starts being a lot of code.

Since an _id has to be a string I'd be ok with saying "The key has to be a string, if you have a complex type do some other SMTs to get to the string you want first" - which would also cover the warning I was talking about [here](https://github.com/IBM/cloudant-kafka-connector/pull/82#discussion_r919900101). I think that is a reasonable limitation.